packages/smart-contracts/test/contracts/InvoiceNFT.test.ts

```diff
@@ -0,0 +1,152 @@
+import '@nomiclabs/hardhat-ethers';
+import { Signer } from 'ethers';
+import { ethers } from 'hardhat';
+import { expect } from 'chai';
+import { TestERC20__factory, TestERC20, InvoiceNFT__factory, InvoiceNFT } from '../../src/types';
+import { BigNumber as BN } from 'ethers';
+
+const BASE_DECIMAL = BN.from(10).pow(BN.from(18));
+
+describe('contract: InvoiceNFT', () => {
+  let deployer: Signer;
+  let user1: Signer, user1Addr: string;
+  let user2: Signer, user2Addr: string;
+  let user3: Signer, user3Addr: string;
+
+  let testToken: TestERC20, nft: InvoiceNFT;
+
+  before(async function () {
+    [deployer, user1, user2, user3] = await ethers.getSigners();
+    user1Addr = await user1.getAddress();
+    user2Addr = await user2.getAddress();
+    user3Addr = await user3.getAddress();
+  });
+
+  beforeEach(async function () {
+    testToken = await new TestERC20__factory(deployer).deploy(BN.from(1000000).mul(BASE_DECIMAL));
+    nft = await new InvoiceNFT__factory(deployer).deploy('Invoice NFT', 'INFT');
+
+    const mintRole = await nft.MINTER_ROLE();
+    await nft.grantRole(mintRole, await deployer.getAddress());
+
+    await testToken.approve(nft.address, ethers.constants.MaxUint256);
+  });
+
+  async function verifyNFTs(userAddr: string, nftIds: any) {
+    const ids = await nft.getIds(userAddr);
+    expect(ids.toString()).equals(nftIds.toString());
+  }
+
+  async function verifyNFTIndexes(indexes: any) {
+    const res = [];
+    for (let i = 1; i <= 5; i++) {
+      res[i - 1] = await nft.getTokenIdIndex(i);
+    }
+    expect(res.toString()).equals(indexes.toString());
+  }
+
+  describe('mint', async function () {
+    it('revert without permission', async function () {
+      await expect(nft.connect(user1).mint(user1Addr, 1, testToken.address, '0x01')).to.be.reverted;
```

Comment:
All the reverts should precise the expected error, with `revertedWithCustomError`

Reply:
Will change. It looks like`revertedWithCustomError` is only available in `@nomicfoundations/hardhat-chai-matchers`, which is incompatible with the existing `hardhat-waffle` and `ethereum-waffle`. I'll clean up the code and follow this migration guide: https://hardhat.org/hardhat-runner/docs/guides/migrating-from-hardhat-waffle